pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin.go

```diff
@@ -308,52 +304,135 @@ func configureListener(
+		SourceService:      proxy.Dataplane.Spec.GetIdentifyingService(),
+		SourceIP:           proxy.Dataplane.GetIP(),
+		DestinationService: r.ResourceOrigin.Name,
+		Mesh:               proxy.Dataplane.GetMeta().GetMesh(),
+		TrafficDirection:   envoy.TrafficDirectionOutbound,
+	}
+
+	backends := pointer.Deref(rctx.Conf().Backends)
+
+	routesConfs, err := buildRoutesMap(listener, rctx)
+	if err != nil {
+		return err
+	}
+
+	routesIds := slices.SortedStableFunc(maps.Keys(routesConfs), kri.Compare)
+
+	hasAtLeastOneBackend := len(util_slices.Filter(util_maps.AllValues(routesConfs), func(conf api.Conf) bool {
+		return len(pointer.Deref(conf.Backends)) > 0
+	})) > 0
+
+	builderForSharedBackend := func(b api.Backend) *Builder[envoy_accesslog.AccessLog] {
+		return BaseAccessLogBuilder(b, defaultFormat, backendsAcc, kumaValues, accessLogSocketPath).
+			ConfigureIf(r.Protocol.IsHTTPBased(), func() Configurer[envoy_accesslog.AccessLog] {
+				return bldrs_accesslog.MetadataFilter(true, bldrs_matcher.NewMetadataBuilder().
+					Configure(bldrs_matcher.Key(namespace, routeMetadataKey)).
+					Configure(bldrs_matcher.NullValue()))
+			})
+	}
+
+	builderForRouteBackend := func(routeID kri.Identifier) func(b api.Backend) *Builder[envoy_accesslog.AccessLog] {
+		return func(b api.Backend) *Builder[envoy_accesslog.AccessLog] {
+			return BaseAccessLogBuilder(b, defaultFormat, backendsAcc, kumaValues, accessLogSocketPath).
+				Configure(bldrs_accesslog.MetadataFilter(false, bldrs_matcher.NewMetadataBuilder().
+					Configure(bldrs_matcher.Key(namespace, routeMetadataKey)).
+					Configure(bldrs_matcher.ExactValue(routeID.String()))))
 		}
+	}
 
-		for _, chain := range listener.FilterChains {
-			if err := listeners_v3.UpdateHTTPConnectionManager(chain, func(hcm *envoy_hcm.HttpConnectionManager) error {
-				hcm.AccessLog = append(hcm.AccessLog, accessLog)
-				return nil
-			}); err != nil {
-				return err
-			}
-			if err := listeners_v3.UpdateTCPProxy(chain, func(tcpProxy *envoy_tcp.TcpProxy) error {
-				tcpProxy.AccessLog = append(tcpProxy.AccessLog, accessLog)
-				return nil
-			}); err != nil {
-				return err
-			}
+	return NewModifier(listener).
+		Configure(bldrs_listener.AccessLogs(util_slices.Map(backends, builderForSharedBackend))).
+		Configure(bldrs_listener.Routes(
+			util_maps.MapValues(
+				routesConfs,
+				func(id kri.Identifier, _ api.Conf) Configurer[routev3.Route] {
+					return bldrs_route.Metadata(routeMetadataKey, id.String())
+				}))).
+		Configure(bldrs_listener.AccessLogs(
+			util_slices.FlatMap(
+				routesIds,
+				func(id kri.Identifier) []*Builder[envoy_accesslog.AccessLog] {
+					return util_slices.Map(
+						pointer.Deref(routesConfs[id].Backends),
+						builderForRouteBackend(id),
+					)
+				}))).
+		ConfigureIf(hasAtLeastOneBackend, func() Configurer[envoy_listener.Listener] {
+			return bldrs_listener.HCM(hcm.LuaFilterAddFirst(setFilterMetadataAsDynamicLuaFilter(namespace, routeMetadataKey)))
```

Comment:
Can we use the `SetMetadata`  filter here? https://www.envoyproxy.io/docs/envoy/v1.30.1/configuration/http/http_filters/set_metadata_filter

If it does the same job, I believe `SetMetadata` is more native.